types/coin.go

```diff
@@ -390,6 +390,52 @@ func (coins Coins) SafeSub(coinsB Coins) (Coins, bool) {
 	return diff, diff.IsAnyNegative()
 }
 
+// Max returns the maximum of each denom of its inputs.
+// The inputs should be sorted. Note that the max might
+// not be equal to either of its inputs. Uses multiset
+// semantics, so unmentioned denoms are implicitly zero.
+func (coins Coins) Max(coinsB Coins) Coins {
+	// coins + coinsB == min(coins, coinsB) + max(coins, coinsB)
+	return coins.Add(coinsB...).Sub(coins.Min(coinsB))
```

Comment:
I don't expect `Max()` to be used as much as `Min()`, but sure.

Reply:
I feel like optimizing for readability is way better right now than performance for this operation as well

Reply:
I had to think more to confirm that this works. With short for loop it would be easier.

Reply:
I had updated it to loop over the `Coin`s. `Max()` is a little trickier than `Min()` since you need to iterate over any leftovers once the overlap has been processed.

I thought about getting fancy and writing a `zip()` helper to iterate over two lists at once, calling processing function to handle the elements. Then `Min`, `Max`, `Add`, and `Sub` could use this common iterator. It could even check for out-of-order or duplicate entries without much overhead. But it's a little harder to get the substructure sharing behavior of an explicit loop.